katana/katana.py

```diff
@@ -90,6 +93,207 @@ def readconfig():
     cfg = json.loads("".join(lines))
     return cfg
 
+@route('/searchkw', method='POST')
+def searchkw():
+    """
+    This method returns the ActionFile Path of the selected Driver.
+    """
+    value = parseString("".join(request.body))
+    indented_xml = "".join(value.toprettyxml(newl='\n'))
+    corrected_xml = remove_extra_newlines_char_xml(indented_xml)
+    corrected_xml = corrected_xml
+    with open('obj.txt', 'w') as f:
+        f.write(corrected_xml)
+    tree = ET.parse('obj.txt')
+    root = tree.getroot()
+    driver_name = root.text
+    global gpysrcdir
+    driver_file = gpysrcdir + '/ProductDrivers/' + driver_name + ".py"
+    actiondir_new = mkactiondirs(driver_file)
+    py_files = mkactionpyfiles(actiondir_new)
+    return py_files
+
+
+@route('/parsexmlobj', method='POST')
+def parsexmlobj():
+    """
+    This method fetch the XML object from the Katana UI of Keyword sequencing tool screen.
+    And parse it to form a new Wrapper keyword & place it in the user provided file path.
+    """
+    method_names = []
+    arg_name_list = ""
+    object_list_new = []
+    class_list_new = []
+    import_action_list = []
+    sample = []
+    py_files = ""
+    doc_string_value = []
+    kw_list_1 = []
+    xmlobj = parseString("".join(request.body))
+    indented_xml = "".join(xmlobj.toprettyxml(newl='\n'))
+    corrected_xml = remove_extra_newlines_char_xml(indented_xml)
+    with open('output.txt', 'w') as f:
+        f.write(corrected_xml)
+    tree = ET.parse('output.txt')
+    root = tree.getroot()
+    WrapperName = getChildTextbyParentTag('output.txt',
+                                          'Details',
+                                          'WrapperName')
+
+    ActionFile = getChildTextbyParentTag('output.txt', 'Details', 'ActionFile')
+    Description = getChildTextbyParentTag('output.txt', 'Details',
+                                          'Description')
+
+    # To check if the user provided wrapper keyword name already exists in action file
+    ActionFile = ActionFile.strip()
+    with open(ActionFile, 'r') as files:
+        for line in files:
```

Comment:
files.readlines() can replace the for loop and the split under

Reply:
Hi @kaheichan 
The purpose for these lines(149 to 154) are to check the user provided wrapper kw name is already exists in user provided action file while saving the wrapper kw.  And here we are searching for the word "def" in each line by splitting it with split(). But the purpose of files.readlines() is reads in the whole file at once and splits it by line(split('\n')).  So I have used "split()" instead of "files.readlines()". Referred this link https://www.peterbe.com/plog/blogitem-040312-1 for understanding it.

Reply:
in such case
To do it in pure text file way
please use if line.strip().startswith("def ") to check if it is actually a def before splitting it into a list of string
or use if line[0] == "def"
and then extract the method name

Or as I mentioned before
Please refer to
https://docs.python.org/2/library/ast.html
https://greentreesnakes.readthedocs.io/en/latest/nodes.html#function-and-class-definitions
to find function def directly

Reply:
Sure @kaheichan . As you said,  I have used line.strip().startswith("def ") to do the operation. Thank you.

Reply:
Done